artifact/v3/version/version.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	int version {-1};
+	string format {"unknown"};
+
+
+	auto err = io::Copy(bw, reader);
+	if (err) {
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, "Failed to parse the version JSON");
+	}
+
+	const json::Json version_json = expected_json.value();
+
+	if (!version_json.IsObject()) {
+		return MakeError(
+			FormatError,
+			"Failed to unmarshal the version from the version JSON data. The received data does not contain a single top-level object");
+	}
+
+	const json::ExpectedJson version_value = version_json.Get("version");
+	if (!version_value) {
+		return MakeError(FormatError, "No version field found");
+	}
+
+	const json::ExpectedInt value_version = version_value.value().GetInt();
+	if (value_version) {
```

Comment:
This logic is wrong now. If the `version` field is not an int, the code continues, but it should fail.

Reply:
Oh

My

God....